phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/GroupedAggregateRegionObserver.java

```diff
@@ -436,36 +622,79 @@ public boolean next(List<Cell> resultsToReturn) throws IOException {
+                            ImmutableBytesPtr originalRowKey = new ImmutableBytesPtr();
+                            result.getKey(originalRowKey);
+                            currentRowKey = originalRowKey;
                             Aggregator[] rowAggregators = groupByCache.cache(key);
+                            groupByCache.cacheAggregateRowKey(key, originalRowKey);
                             // Aggregate values here
                             aggregators.aggregate(rowAggregators, result);
                         }
                         now = EnvironmentEdgeManager.currentTimeMillis();
-                    } while (hasMore && groupByCache.size() < limit && (now - startTime) < pageSizeMs);
-                    if (hasMore && groupByCache.size() < limit && (now - startTime) >= pageSizeMs) {
-                        // Return a dummy result as we have processed a page worth of rows
-                        // but we are not ready to aggregate
-                        getDummyResult(resultsToReturn);
-                        return true;
-                    }
+                        if (hasMore && groupByCache.size() < limit &&
+                                (now - startTime) >= pageSizeMs) {
+                            return getDummyResult(resultsToReturn);
+                        }
+                    } while (hasMore && groupByCache.size() < limit);
                     regionScanner = groupByCache.getScanner(delegate);
                     // Do not sort here, but sort back on the client instead
                     // The reason is that if the scan ever extends beyond a region
                     // (which can happen if we're basing our parallelization split
                     // points on old metadata), we'll get incorrect query results.
                     return regionScanner.next(resultsToReturn);
                 }
+            } catch (Exception e) {
+                LOGGER.error("Unordered group-by scanner next encountered error.", e);
+                if (e instanceof IOException) {
+                    throw e;
+                } else {
+                    throw new IOException(e);
+                }
             } finally {
                 if (acquiredLock) region.closeRegionOperation();
             }
         }
 
+        /**
+         * Retrieve dummy rowkey and return to the client.
+         *
+         * @param resultsToReturn dummy cell.
+         * @return true if more rows exist after this one, false if scanner is done.
+         */
+        private boolean getDummyResult(List<Cell> resultsToReturn) {
+            if (lastReturnedRowKey != null) {
+                ScanUtil.getDummyResult(lastReturnedRowKey, resultsToReturn);
+                return true;
+            }
+            if (scanStartRowKey.length > 0 && !ScanUtil.isLocalIndex(scan)) {
+                if (Bytes.compareTo(prevScanStartRowKey, scanStartRowKey) !=
```

Comment:
This is the same condition we used above on line 493..495 for detecting region move Maybe introduce a util function in this class. That will also make it self descriptive that we are implementing this logic to handle region move